connectors/sources/confluence.py

```diff
@@ -256,6 +323,214 @@ async def ping(self):
+            # base64 on macOS will add a EOL, so we strip() here
+            document["_attachment"] = (await target_file.read()).strip()
+        await remove(source_file_name)
+        logger.debug(f"Downloaded {attachment_name} for {attachment_size} bytes ")
+        return document
+
+    async def _attachment_coro(self, document):
+        """Coroutine to add attachments to Queue and download content
+
+        Args:
+            document (dict): Formatted document of page/blogpost
+        """
+        async for attachment, download_link in self.fetch_attachments(
+            content_id=document["_id"],
+            parent_name=document["title"],
+            parent_space=document["space"],
+            parent_type=document["type"],
+        ):
+            await self.queue.put(
+                (  # pyright: ignore
+                    attachment,
+                    partial(
+                        self.download_attachment,
+                        download_link[1:],
+                        copy(attachment),
+                    ),
+                )
+            )
+        await self.queue.put("FINISHED_TASK")  # pyright: ignore
+
+    async def _space_coro(self):
+        """Coroutine to add spaces documents to Queue"""
+        async for space in self.get_spaces():
+            await self.queue.put((space, None))  # pyright: ignore
+        await self.queue.put("FINISHED_TASK")  # pyright: ignore
+
+    async def _page_blog_coro(self, api_query):
+        """Coroutine to add pages/blogposts to Queue
+
+        Args:
+            api_query (str): API Query Parameters for fetching page/blogpost
+        """
+        async for document, attachment_count in self.fetch_documents(api_query):
+            await self.queue.put((document, None))  # pyright: ignore
+            if attachment_count > 0:
+                await self.fetchers.put(partial(self._attachment_coro, copy(document)))
+                self.fetcher_count += 1
+        await self.queue.put("FINISHED_TASK")  # pyright: ignore
+
+    async def _consumer(self):
+        """Async generator to process entries of the queue
+
+        Yields:
+            dictionary: Documents from Confluence.
+        """
+        while self.fetcher_count > 0:
+            _, item = await self.queue.get()
+            if item == "FINISHED_TASK":
+                self.fetcher_count -= 1
+            else:
+                yield item
+
     async def get_docs(self, filtering=None):
-        # yield dummy document to make the chunked PR work, subsequent PR will replace it with actual implementation
-        yield {"_id": "123", "timestamp": iso_utc()}, None
+        """Executes the logic to fetch Confluence content in async manner.
+
+        Args:
+            filtering (Filtering): Object of Class Filtering
+        Yields:
+            dictionary: dictionary containing meta-data of the content.
+        """
+        await self.fetchers.put(self._space_coro)
```

Comment:
A similar mechanism is already present in the framework. Is there a specific reason why we don't yield lazy download co-routines to the framework and let the framework handle it?

Reply:
`lazy_download` used in the framework is responsible for downloading content of objects that were yielded. In Confluence, content is only present in the `attachment` object. 
But, the fetching of  `attachments` needs to happen after `pages` and `blogs`  are fetched to use their ids as reference for api call. This dependency slows down the connector if we don't allow the connector to start grabbing attachments for the already fetched pages, blogs. 

Using the ConcurrentTasks directly here improves the performance of the connector, as it allows fetching all objects in parallel. 
Downloading the content of `attachment` is still happening via `lazy_download` by the framework.